Templating support for zkg

ckreibich opened the pull request:
This is a more fully baked version of the templating support we've previously discussed over in #93. The idea is as follows:

- `zkg` gains support for instantiating new Zeek packages from a package template.
- These templates are independent git repositories, with corresponding support for versioning, branches, etc.
- Templates provide two kinds of functionality: Zeek packages and optional additional features, such as CI, plugins, Spicy support, specific copyrights, etc. You can mix-and-match these into your resulting package.
- Templates support named parameters that get substituted upon instantiation. The user provides these parameters as user vars (via user prompting, environment variables, previously stored values in the zkg config, or `--user-var` -- more on this below). The template expands the given user variables into its required parameters. This is an n:m relationship -- for example, given a package name, the template might define one parameter with that name as given, one sanitized and upper-cased for use in a module string, one derived from a combination of user vars, etc. User vars are the external interface, parameters their internal application.
- Templates are mainly a tree of file content, but driven by Python code (`__init__.py` at the toplevel). This enables flexibility with input validation, how to instantiate content, deal with potential conflicts across features (overwriting vs selectively editing files like zkg.meta, for example), or modernizing content (say via patching, moving files around, renaming things, overwriting them, sanity-checking existing content, etc).
- We maintain an official default template at https://github.com/zeek/package-template, but users can define and use their own.
- The target audience for template development is primarily us Zeek devs, the audience for template use is anyone who builds Zeek packages.

What does this address?

- A more accessible package creation mechanism. `init-plugin` provides a solid layout but resides only in the source tree.
- A more flexible package creation mechanism: pick just the package features you need, with many at your disposal.
- A more explicit default source of truth for what Zeek packages should look like.
- Assistance (with caveats) with keeping package content modern over time.

What is not there yet?

- An agreed-upon v1 template. https://github.com/zeek/package-template exists and works (providing plugins and really basic Github CI as starting points), but people will have thoughts/opinions about what should go in there. I suggest we label this templating feature beta to allow ourselves time to iterate on the template. I'm definitely open to suggestions here.

- Modernization of an existing template-derived package. This is a natural next step but the PR is already on the large side, so I figured I'd wait. (I think such modernizations will likely end up being a mix of doable approaches when a template clearly owns content, and suggestions/guidance for modernization when things are less clear, e.g. via conflicting patches. Regardless, I do think they will be useful for the user.)

- A command that prints information about a package (see below). This should be there because it's otherwise difficult for the user to know what features a package supports, what versions it offers, etc. This is much like `zkg info` for packages.

- Support for combining multiple templates. For example, one might want to use our default template, but overlay a CI feature for a specific platform. I feel this would often be valuable, and simpler to maintain than forking the standard one and maintaining additional features on top of it.

Other things in this PR:

- I added two things to zkg's user-var support: you can now enter user vars via the command line, using `--user-var NAME=VAL`. This works for any zkg command that currently prompts for user-var input. I've also tweaked the way user input works: the default or pre-recorded value of a user variable now gets populated into the input prompt. This maintains the previous behavior of letting you "just hit enter" to confirm the existing value, but also lets you edit the value, including setting the variable to an empty string.

- I noticed that on my systems the btest output got cluttered with git warnings about the lack of a configured default branch name (`init.defaultBranch`), which newer git versions request. I added this to the `scripts/git` wrapper.

If you'd like to give this a go:

- Install this zkg as usual, then try e.g. `zkg create --packagedir foobar` or `zkg create --packagedir foobar --feature plugin github-ci --user-var name=FooBar --user-var namespace=MyOrg`.

You can clone https://github.com/zeek/package-template to your system and point `zkg create` at it via the --template argument if you'd like to tweak the template.

Things I'm unsure about:

- Command naming. "zkg create" is concise, but it seems likely we'll want a `zkg template` command set, e.g. to have `zkg template info`, `zkg template update`, etc. So template instantiation could live there from the outset, perhaps as `zkg template build`.

- What happens to `zeek-aux/plugin-support`? My suggestion would be to reduce it to what its name suggests: plugin support. We want to continue standalone plugins, for which this seems a good fit.

Sorry for the size of the whole thing. :grimacing: 

Resolves #93.

ckreibich committed:
Sort imported modules alphabetically, no other change

ckreibich committed:
Migrate some git helpers to _util for broader access

ckreibich committed:
Refactor user vars into their own class and module

- Moves user-var functionality to a new uservar module and adds
UserVar, a class representing an individual user variable. These can
be created explicitly, parsed from a dict, or parsed from a NAME=VAL
string.

- Adds support for --user-arg NAME=VAL for the install, unbundle, and
upgrade commands. This can be provided repeatedly and lets you provide
user variables via the command line.

- Slightly tweaks the user prompt to accommodate situations other than
package installation.

ckreibich committed:
Update the user-mode test to reflect template addition

ckreibich committed:
Prefill the prompt for user vars with current value

This preserves the current model of allowing a quick hit of enter to
confirm the existing value, but also lets you empty out the value to
provide an empty string as the new value.

ckreibich committed:
Add btest for providing user var via command-line arguments

ckreibich committed:
Add _util.load_source() for sourcing a Python script as a module

Templates are driven by a toplevel __init__.py script, which we source
via this function.

ckreibich committed:
Add Zeek package templating module

This provides infrastructure for structuring and instantiating Zeek
package templates. Templates are standalone git repos, managed as
follows:

- Template objects define package templates, including an output
directory, a specific package template, any additional features, and
parameters substituted when instantiating. Parameters derive from user
variables in a way defined by the template.

- The Package and Feature classes model template content and allow
validation of input parameters and instantiation of content. The
package class has additional functionality for initializing new
packages as git repos and adding template metadata to its zkg.meta.

- A hierarchy of exceptions accommodates template-specific problems.

These aren't currently imported into the zeekpkg namespace -- you need
to use zeekpkg.template.* when accessing those components.

ckreibich committed:
Add a "slug" helper for user vars

For templates it'll often be handy to render a variable value into a
file-system-safe identifier ("slug"). This provides uservar.slugify()
for this purpose.

ckreibich committed:
Add templating functionality to zkg toplevel script

- Provide "zkg create" to instantiate a template, with optional
features and user vars for parameterization.

- ZKG_DEFAULT_TEMPLATE points at the default template repo,
https://github.com/zeek/package-template.

ckreibich committed:
Add btests for template instantiation

ckreibich committed:
Provide default branch name in our git wrapper

Newer git versions trigger a warning when the default branch name is
not set, which cluttered up the logs. This adds the default name,
preserving "master" since that branch name is currently visible in
test baselines.

ckreibich commented:
I realized I'd like to make a tweak to allow user vars to be tied to specific features instead of the template as a whole — that way, when a feature requires user input that input will only happen when template instantiation actually uses that feature.

Also, there's no documentation yet. I'll be happy to work on that as this moves forward.

jsiwek commented on zeekpkg/package.py:
```diff
@@ -85,42 +86,6 @@ def short_description(metadata_dict):
     return rval.lstrip()
 
 
-def user_vars(metadata_dict):
```
Don't know how likely it is someone relies on this API, but since it's not originally marked as internal in any way, I'd just leave it be an alias for `UserVar.parse_dict`.

ckreibich replied:
Whoops! Yes, thanks for catching this.

ckreibich replied:
Agreed, yeah. Will do.

jsiwek commented on zeekpkg/manager.py:
```diff
@@ -2289,7 +2292,18 @@ def _stage(self, package, version, clone,
         raw_metadata = _get_package_metadata(raw_metadata_parser)
-        requested_user_vars = user_vars(raw_metadata)
+
+        # Ensure any listed executables exist as advertised. Do this first so
+        # that we don't build anything if it fails.
+        for p in self._get_executables(raw_metadata):
+            full_path = os.path.join(clone.working_dir, p)
+            if not os.path.isfile(full_path):
+                return str.format("executable '{}' is missing", p)
+
+            if not os.access(full_path, os.X_OK):
+                return str.format("file '{}' is not executable", p)
```
```suggestion
```
Guessing this sneaked in while resolving a Git conflict?  (This logic got moved to the end and shouldn't be here anymore).

ckreibich replied:
Whoops! Yes, thanks for catching this. I've committed your suggestion.

jsiwek reviewed (commented):
Looks nice, thanks.  I didn't review or test things comprehensively, but I didn't have any feedback about the basic structure of it.  Not sure if you want to take that a sign to just move forward with anything else you have planned (e.g. documentation) or wait for other feedback.

> * Command naming. "zkg create" is concise, but it seems likely we'll want a `zkg template` command set, e.g. to have `zkg template info`, `zkg template update`, etc. So template instantiation could live there from the outset, perhaps as `zkg template build`.

May not be that critical -- if there's later a family of `zkg template <operation>`, seems like a `zkg create` that gets added now can turn into simple alias/shorthand for something living under `zkg template <operation>`.

> * What happens to `zeek-aux/plugin-support`? My suggestion would be to reduce it to what its name suggests: plugin support. We want to continue standalone plugins, for which this seems a good fit.

Not sure much has to be done with it except maybe try to encourage/advertise this method of package creation in docs and de-emphasize any `init-plugin`, but otherwise `init-plugin` still seems like a way to create a valid plugin template and could be left as is if people want to use it (e.g. if they're less `zkg` or "package" oriented and more into just making a standalone "plugin" for their own use).

ckreibich committed:
Update zeekpkg/manager.py

Co-authored-by: Jon Siwek <jsiwek@corelight.com>

ckreibich committed:
Preserve the zeekpkg.package.user_vars() API

ckreibich commented:
Thanks for looking it over, Jon. I'll write up documentation and put a bit of work into the default template, so it's in decent shape when we merge this. I'll keep rebasing this branch in the meantime as needed. I'm thinking we can merge this next week then, if that sounds good to you.

ckreibich committed:
Add "template info" command

"template info" shows information about the provided template (or
the default), in either JSON or plaintext. This verifies whether
a given template is loadable and reports supported features,
user vars, and tagged versions.

ckreibich committed:
Add btest for "zkg template info"

ckreibich committed:
Add documentation for the templating support

ckreibich commented:
@jsiwek the above push makes minor changes to what you already reviewed:

- It rearranges user variables a bit inside the template content classes. I realized that previously the template couldn't share the value of a single variable across its content, which I imagine will be common (e.g. your name might be used by multiple features).
- There's a first `template` sub-command: `zkg template info` summarizes a template's properties, much like `zkg info` does for packages. I've kept `zkg create` for now since it's concise, but we could also move it over to `zkg template` now.
- Added some resilience to incorrect input for `zkg create`, pylinting, and added more helpful error/warning messages.
- Documentation. I've not yet removed the old package production walkthroughs, just put the templating docs first.

I'm thinking this is a good time to merge, but happy to hear your thoughts ... thanks!

jsiwek commented:
Thanks, just wondering what the thought is on how long to expect this to be treated as a "beta" feature (and e.g. `API_VERSION` of 1.0.0-rc1).  Think a zkg v2.9 release waits on that?

ckreibich commented:
Thanks Jon! Yes exactly — I was thinking that when we release a new zkg version the beta status can go, and the API moves to 1.0.0. I've been meaning to ask you how you version zkg (what would be 2.8.1 vs 2.9, or even 3.0).

The beta and rc are just there because I'm nervous about the fact that mostly I've been playing with this and also developing it, so I think I'm prone to missing things. If people don't report major problems in the next 1-2 weeks, I think these could go.

jsiwek commented:
> I've been meaning to ask you how you version zkg (what would be 2.8.1 vs 2.9, or even 3.0).

Probably what you expect:
* Major-version: significant changes, new way of doing/structuring things, or backwards incompatible breakages
* Minor-version: added new features
* Patch-level: only bug fixes

> If people don't report major problems in the next 1-2 weeks, I think these could go.

Sounds fine.  I can check back in 2 weeks and do a v2.9 if there's no concerns.
